### PR TITLE
ESQL: Fix filtered grouping on ords

### LINKED ISSUE #114897: ESQL: IAE with stats filtering
### Description

The following queries:

`FROM employees | STATS max(salary) WHERE first_name == "" BY job_positions`
`FROM employees | STATS max(salary) WHERE salary > 0 BY job_positions`

fail with:

```
[DEBUG][o.e.x.e.a.EsqlResponseListener] [runTask-0] Request failed with status [BAD_REQUEST]: java.lang.IllegalArgumentException: expected class org.elasticsearch.compute.aggregation.MaxIntGroupingAggregatorFunction; got class org.elasticsearch.compute.aggregation.FilteredGroupingAggregatorFunction
        at org.elasticsearch.compute.aggregation.MaxIntGroupingAggregatorFunction.addIntermediateRowInput(MaxIntGroupingAggregatorFunction.java:186)
        at org.elasticsearch.compute.aggregation.FilteredGroupingAggregatorFunction.addIntermediateRowInput(FilteredGroupingAggregatorFunction.java:100)
        at org.elasticsearch.compute.aggregation.GroupingAggregator.addIntermediateRow(GroupingAggregator.java:65)
        at org.elasticsearch.compute.operator.OrdinalsGroupingOperator.mergeOrdinalsSegmentResults(OrdinalsGroupingOperator.java:251)
        at org.elasticsearch.compute.operator.OrdinalsGroupingOperator.getOutput(OrdinalsGroupingOperator.java:218)
        at org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:258)
        at org.elasticsearch.compute.operator.Driver.run(Driver.java:189)
        at org.elasticsearch.compute.operator.Driver$1.doRun(Driver.java:378)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:34)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:1023)
        at org.elasticsearch.server@9.0.0-SNAPSHOT/org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:27)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
```

### PR DESCRIPTION
**Pull request by nik9000**
This fixes filtered aggs when they are grouped on a field with ordinals. This looks like:
```
| STATS max = max(salary) WHERE salary > 0 BY job_positions
```
when the `job_positions` field is a keyword field with doc values. In that case we use a faster group-by-segment-ordinals algorithm that needs to be able to merge the results of aggregators from multiple segments. This previously failed with a `ClassCastException` because of a mistake.

Also! the group-by-segment-ordinals algorithm wasn't properly releasing the closure used to add inputs, causing a breaker size leak. This wasn't really leaking memory, but leaking *tracking* of memory.

Closes #114897

### CODE REVIEW

**Review by bpintea — APPROVED** [review]
Lgtm.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec [hunk 1]
```diff
@@ -2502,3 +2502,71 @@ FROM employees | eval x = [1,2,3], y = 5 + 6 | stats m = max(y) by y+1
 m:integer | y+1:integer
 11        | 12
 ;
+
+filterIsAlwaysTrue
```
Comment (ivancea) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/stats.csv-spec:2506]: Are those the first CSV tests for filtered aggs? Or are they specifically to test the issue here?

**Review by ivancea — APPROVED** [review]
LGTM!

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/FilteredGroupingAggregatorFunction.java [hunk 2]
```diff
@@ -97,7 +97,7 @@ public void addIntermediateInput(int positionOffset, IntVector groupIdVector, Pa
 
     @Override
     public void addIntermediateRowInput(int groupId, GroupingAggregatorFunction input, int position) {
-        next.addIntermediateRowInput(groupId, input, position);
+        next.addIntermediateRowInput(groupId, ((FilteredGroupingAggregatorFunction) input).next(), position);
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/FilteredGroupingAggregatorFunction.java:100]: There we two bugs with ordinals. This is the ClassCast one. The other was the leak.

**Comment by nik9000** [comment]
run elasticsearch-ci/rest-compatibility

**Comment by nik9000** [comment]
I'm backporting to 8.16.

**Comment by nik9000** [comment]
Backport: 115464

**Pull request merged**